adr/0009-extension-structure.adoc

```diff
@@ -0,0 +1,261 @@
+
+== Decision
+
+Extensions may adopt a standardized structure, composed of one mandatory and several optional modules, depending on their complexity and integration needs:
+
+* `runtime`: Contains the runtime implementation.
+It may expose public APIs if needed but should prefer delegation to the `runtime-api` module (See <<the-place-of-the-public-api>>).
+This module also contains the `recorders` and other runtime logic.
+* `runtime-api`: Contains public APIs intended to be consumed by other extensions or application code.
+This module should have minimal dependencies.
+Depending on an `runtime-api` module does not transitively include the full extension (i.e., it avoids pulling in runtime or deployment logic). See <<coupling-api-and-spi>> for more details.
+* `deployment`: Contains build-time logic, including `BuildSteps`.
+This module may define internal build items.
+* `deployment-api`: Contains build-time APIs intended to be reused across multiple extensions.
+Extensions contributing to or using build items from other extensions should depend on `deployment-api` modules.
+* `runtime-dev`: Contains runtime classes used exclusively in dev mode (e.g., for the Dev UI). This avoids shipping dev-only classes into production artifacts.
+* `codestart`: Contains the _Codestart_ templates for the extension. See https://quarkus.io/guides/extension-codestart[Codestart Guide] for more details.
+* `codegen`: Contains code generation utilities such as code pre-processors invoked at build time.
+
+Only the `runtime` module is mandatory; the others are optional based on the extension's needs.
+
+
+.New structure for Quarkus extensions
+[source,tree]
+----
+my-extension/
+├── runtime/
+├── runtime-api/
+├── deployment/
+├── deployment-api/
+├── runtime-dev/
+├── codegen/
+└── codestart/
+----
+
+=== Additional Rules
+
+* Module dependencies must follow strict rules:
+* `deployment` depends on `runtime`, and `deployment-api` (if defined).
+* `runtime` depends on `api` (if defined).
+* `runtime-dev`, if defined, depends on `runtime`.
+* External consumers should rely only on `api` or `deployment-api` to be loose-coupled (See <<coupling-api-and-spi>>.).
+* External consumers should rely only on `runtime` or `deployment` to be tightly-coupled (forcing the target extension to be present at execution). See <<coupling-api-and-spi>>.
+* Public APIs must be explicitly documented, including build items meant for inter-extension use.
+* Public build items must document their intended usage (produced, consumed, or both).
+* Each module should provide an `Automatic-Module-Name` based on the root package:
+
+	Runtime:: `io.<quarkus|quarkiverse>.<extension-name>`
+	Deployment:: `io.<quarkus|quarkiverse>.<extension-name>.deployment`
+	Runtime API:: `io.<quarkus|quarkiverse>.<extension-name>.api`
+	Deployment API:: `io.<quarkus|quarkiverse>.<extension-name>.deployment.api`
+	Runtime Dev:: `io.<quarkus|quarkiverse>.<extension-name>.runtime.dev`
```

Comment:
Looks like `io.<quarkus|quarkiverse>.<extension-name>.runtime` is always supposed to be an empty package?

If so, should it be `io.<quarkus|quarkiverse>.<extension-name>.dev` instead?